docs/requirements/ADP-957.md

```diff
@@ -0,0 +1,156 @@
+# Basic definitions
+
+*   **Collateral inputs**, in a transaction, are used to cover transaction fees
+    in the case that a non-native script fails. The term **collateral** refers
+    to the total ada contained in the UTxOs referenced by collateral inputs.
+    Collateral inputs can **only** contain ada: they may **not** contain
+    non-ada assets.
+
+*   The protocol parameter `collateralPercent` is used to specify the
+    percentage of the total transaction fee its collateral must (at a minimum)
+    cover. The collateral inputs must not themselves be locked by a script.
+
+*   The protocol parameter `maxCollateralInputs` is used to limit,
+    additionally, the total number of collateral inputs, and thus the total
+    number of additional signatures that must be checked during validation.
+
+# Proposed requirements
+
+1.  The wallet, when requested to create a transaction requiring collateral,
+    should reserve one or more pure-ada UTxO entries to serve as collateral
+    inputs, where the total balance of collateral inputs is greater than or
+    equal to the minimum required, as determined by the `collateralPercent`
+    protocol parameter and the transaction fee.
+
+2.  The wallet, when requested to create a transaction requiring collateral,
+    should fail with an error if the total balance of pure-ada UTxO entries is
+    insufficient to cover the minimum collateral amount, as determined by the
+    `collateralPercent` protocol parameter and the transaction fee.
+
+3.  The wallet's create transaction operation should accept a parameter
+    `collateralPercentMaximum`, with a default value that can be overridden
+    by the caller. The wallet, when requested to create a transaction requiring
+    collateral, should fail with an error if it **is** possible to meet the
+    minimum amount, but **not** possible to find a set of pure-ada entries
+    whose total balance is less than or equal to
+    `collateralPercentMaximum × transactionFee`.
+
+4.  The wallet should ensure that any UTxO entries used for collateral are
+    **locked** until the transactions that include them are validated.
+
+5.  The wallet should report the balance of posted collateral entries as part
+    of the total balance, but not as part of the available balance.
```

Comment:
If the tx is valid, will the collateral be spent like normal inputs? If so, is requirement not just how the wallet currently works?

And additionally if so, it seems misleading: _Inputs_ of pending txs are excluded from both the available and total balance. It's the corresponding _change outputs_ that gets added to the total balance only.

Reply:
> If the tx is valid, will the collateral be spent like normal inputs? If so, is requirement not just how the wallet currently works?

My understanding is that:

If a transaction in a block is marked as having **_passed validation_**, then we must apply the following state transitions:

```
UTxO       := UTxO ⋪ inputs tx
pendingTxs := pendingTxs − tx
```

If a transaction in a block is marked as having **_failed validation_**, then we must apply the following state transitions:

```
UTxO       := UTxO ⋪ collateralInputs tx
pendingTxs := pendingTxs − tx
```

So in other words, when we process a transaction from a block, we either remove the collateral inputs from the UTxO, or we remove the ordinary inputs from the UTxO, but not both.

Reply:
@Anviking the treatment of transactions in blocks is now covered in the requirements. Hopefully this should be clear, but let me know if you think anything needs to be clarified or improved further!